runtime/foucoco/src/lib.rs

```diff
@@ -921,6 +924,341 @@ parameter_types! {
 	pub Schedule: pallet_contracts::Schedule<Runtime> = Default::default();
 }
 
+use frame_support::{
+	log::{error, warn},
+	pallet_prelude::*,
+};
+use sp_std::vec::Vec;
+
+use pallet_contracts::chain_extension::{
+	ChainExtension,
+	Environment,
+	Ext,
+	InitState,
+	RetVal,
+	SysConfig,
+	// UncheckedFrom,
+};
+use sp_core::crypto::UncheckedFrom;
+
+#[derive(Default)]
+pub struct Psp22Extension;
+
+use runtime_common::chain_ext::*;
+pub(crate) type BalanceOfForChainExt<T> =
+	<<T as orml_currencies::Config>::MultiCurrency as orml_traits::MultiCurrency<
+		<T as frame_system::Config>::AccountId,
+	>>::Balance;
+
+impl<T> ChainExtension<T> for Psp22Extension
+where
+	T: SysConfig
+		+ orml_tokens::Config<CurrencyId = CurrencyId>
+		+ pallet_contracts::Config
+		+ orml_currencies::Config<MultiCurrency = Tokens, AccountId = AccountId>
+		+ orml_currencies_allowance_ext::Config,
+	<T as SysConfig>::AccountId: UncheckedFrom<<T as SysConfig>::Hash> + AsRef<[u8]>,
+{
+	fn call<E: Ext>(&mut self, mut env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+	where
+		E: Ext<T = T>,
+		<E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+	{
+		let func_id = env.func_id();
+
+		warn!("func_id : {}", func_id);
+
+		match func_id {
+			//transfer
+			1105 => {
+				let ext = env.ext();
+				let address = ext.address().clone();
+				let caller = ext.caller().clone();
+				let mut env = env.buf_in_buf_out();
+				let create_asset: (
+					OriginType,
+					u8,
+					[u8; 12],
+					[u8; 32],
```

Comment:
Let's create type aliases for these three things so that this assignment gets easier to understand.

Reply:
Done